apis/v1alpha2/gatewayclass_types.go

```diff
@@ -0,0 +1,205 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api,scope=Cluster,shortName=gc
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Controller",type=string,JSONPath=`.spec.controller`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:printcolumn:name="Description",type=string,JSONPath=`.spec.description`,priority=1
+
+// GatewayClass describes a class of Gateways available to the user
+// for creating Gateway resources.
+//
+// GatewayClass is a Cluster level resource.
+type GatewayClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of GatewayClass.
+	Spec GatewayClassSpec `json:"spec"`
+
+	// Status defines the current state of GatewayClass.
+	//
+	// +kubebuilder:default={conditions: {{type: "Admitted", status: "False", message: "Waiting for controller", reason: "Waiting", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status GatewayClassStatus `json:"status,omitempty"`
+}
+
+// GatewayClassSpec reflects the configuration of a class of Gateways.
+type GatewayClassSpec struct {
+	// Controller is a domain/path string that indicates the
+	// controller that is managing Gateways of this class.
+	//
+	// Example: "acme.io/gateway-controller".
+	//
+	// This field is not mutable and cannot be empty.
+	//
+	// The format of this field is DOMAIN "/" PATH, where DOMAIN is a valid
+	// Kubernetes name (https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names)
+	// and PATH is a valid HTTP path as defined by RFC 3986.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MaxLength=253
+	Controller string `json:"controller"`
+
+	// ParametersRef is a reference to a resource that contains the configuration
+	// parameters corresponding to the GatewayClass. This is optional if the
+	// controller does not require any additional configuration.
+	//
+	// ParametersRef can reference a standard Kubernetes resource, i.e. ConfigMap,
+	// or an implementation-specific custom resource. The resource can be
+	// cluster-scoped or namespace-scoped.
+	//
+	// If the referent cannot be found, the GatewayClass's "InvalidParameters"
+	// status condition will be true.
+	//
+	// Support: Custom
+	//
+	// +optional
+	ParametersRef *ParametersReference `json:"parametersRef,omitempty"`
+
+	// Description helps describe a GatewayClass with more details.
+	//
+	//
+	// +kubebuilder:validation:MaxLength=64
```

Comment:
more details in 64 char?

Reply:
Yeah this could likely be longer. We wanted it to be something that could at least theoretically be displayed with kubectl wide table output which limits the possible length. Most decisions around validation length have so far been taken with the idea that we can increase minimums in the future much easier than we can decrease them.

Reply:
Is there any similar precedent?

Reply:
cc @jpeach

Reply:
>Is there any similar precedent?

Not that I know of. It would of course be trivial to port back to IngressClass if there were interest.